pkg/rpc/request/types.go

```diff
@@ -35,38 +40,102 @@ type Raw struct {
 	ID        int           `json:"id"`
 }
 
+// Request contains standard JSON-RPC 2.0 request and batch of
+// requests: http://www.jsonrpc.org/specification.
+// It's used in server to represent incoming queries.
+type Request struct {
+	In    *In
+	Batch Batch
+}
+
 // In represents a standard JSON-RPC 2.0
-// request: http://www.jsonrpc.org/specification#request_object. It's used in
-// server to represent incoming queries.
+// request: http://www.jsonrpc.org/specification#request_object.
 type In struct {
 	JSONRPC   string          `json:"jsonrpc"`
 	Method    string          `json:"method"`
 	RawParams json.RawMessage `json:"params,omitempty"`
 	RawID     json.RawMessage `json:"id,omitempty"`
+	Error     error           `json:"-"`
 }
 
-// NewIn creates a new Request struct.
-func NewIn() *In {
-	return &In{
-		JSONRPC: JSONRPCVersion,
+// Batch represents a standard JSON-RPC 2.0
+// batch: https://www.jsonrpc.org/specification#batch.
+type Batch []In
+
+// MarshalJSON implements json.Marshaler interface
+func (r Request) MarshalJSON() ([]byte, error) {
+	if r.In != nil {
+		return json.Marshal(r.In)
+	}
+	return json.Marshal(r.Batch)
+}
+
+// UnmarshalJSON implements json.Unmarshaler interface.
+func (r *Request) UnmarshalJSON(data []byte) error {
+	var (
+		in    *In
+		batch Batch
+	)
+	in = NewIn()
+	err := json.Unmarshal(data, in)
+	if err == nil {
+		if in.JSONRPC != JSONRPCVersion {
+			return fmt.Errorf("invalid version, expected 2.0 got: '%s'", in.JSONRPC)
+		}
+		r.In = in
+		return nil
+	}
+	decoder := json.NewDecoder(bytes.NewReader(data))
+	_, err = decoder.Token() // read `[`
+	if err != nil {
+		return err
 	}
+	count := 0
+	for decoder.More() {
+		if count > maxBatchSize {
+			return fmt.Errorf("the number of requests in batch shouldn't exceed %d", maxBatchSize)
+		}
+		in = NewIn()
+		decodeErr := decoder.Decode(in)
+		if decodeErr != nil {
+			in.Error = decodeErr
```

Comment:
Fixed, but I still think that `Error` field is useful for us, e.g. when one of requests from batch represents correct JSON, but method is wrong or parameters are bad. Can we keep `Error` field for such cases?

Reply:
But you'll create a response structure for them anyway and it'll contain this error, is it really a problem?

Reply:
The only place you're using it now is
```
 			in.Error = fmt.Errorf("invalid version, expected 2.0 got: '%s'", in.JSONRPC)
```
and it can be solved by moving this logic out of unmarshaller code. Let request handler check for it and create a proper error response.

Reply:
Yes, you're right, fixed.